### ISSUE TITLE
Unexpected error: HexaPDF::MalformedPDFError

### ISSUE THREAD

**Issue by efatsi** [gettalong/hexapdf#51]
I'm getting the following error when trying to retrieve the `pages` of a local PDF:

```
HexaPDF::MalformedPDFError: PDF malformed around position : The oid,gen (43,0) values of the indirect object don't match the values (44,0) from the xref
```

The code:
```rb
pdf = HexaPDF::Document.open("/Users/efatsi/Desktop/Ratified.pdf")
pdf.pages.count # this line triggers the error.
```

The PDF appears mostly normal (desktop icon is slightly different), and opens well enough in other tools, so I'm not sure what this error means. The only other clue I have is that if I open the PDF in Preview, and then export it as a PDF, then that version works. Meaning OSX Preview is able to squash whatever formatting issue there is. 

Curious if you had any thoughts as to what might be going on here. Unfortunately I can't share the culprit PDF as it contains sensitive information, but I could try to track down a similarly messed up one if that would help diagnose.

**Comment by efatsi** [comment]
HexaPDF part of the backtrace of the error:
```
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/parser.rb:334:in `raise_malformed'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/parser.rb:78:in `load_object'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/revisions.rb:64:in `block in from_io'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/revision.rb:196:in `load_object'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/revision.rb:109:in `object'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/document.rb:156:in `object'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/document.rb:167:in `deref'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/dictionary.rb:155:in `[]'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/type/trailer.rb:70:in `catalog'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/document.rb:493:in `catalog'
/Users/efatsi/.asdf/installs/ruby/2.5.1/lib/ruby/gems/2.5.0/gems/hexapdf-0.7.0/lib/hexapdf/document/pages.rb:135:in `each'
```

**Comment by gettalong** [comment]
As you deduced, it seems that the PDF is malformed. Most viewers work around this kind of malformedness by discarding the XREF table and parsing the whole PDF file top to bottom. HexaPDF doesn't currently do this (although it tries to be forgiving for other kinds of problems).

This means that currently the only way forward would be to fix the broken PDF before handing it off to HexaPDF.

**Comment by matpowel** [comment]
FWIW @gettalong we're hitting this problem semi-regularly, seems quite common in the wild. Is it a big change to parse the document and ignore the xref in this case? Or an easier workaround? Would your recommendation to "fix" it be to load and save it back out using something else like VIPS? Thanks!

**Comment by gettalong** [comment]
@matpowel For now I want to finish everything and get the next release out. If you have time it would help if you could provide some failing PDFs for inspection to see what kind of underlying problems lead to this error (or to other similar ones). I can prioritize this for the release after this one, if that helps.

My recommendation is to use a tool that completely understands PDF, don't know if this is the case with VIPS. Have a look at https://superuser.com/questions/278562/how-can-i-fix-repair-a-corrupted-pdf-file where mutool is mentioned. qpdf should also work.

**Comment by efatsi** [comment]
Thanks @gettalong, unfortunately I'm not in a position to test this out at the moment, I never had direct access to the malformed PDFs, only complaints from the software users and error logs.

**Comment by gettalong** [comment]
@efatsi Thanks for the response and no problem!

**Comment by pashamur** [comment]
I tried out the devel branch against one such corrupted PDF and got `Problem encountered: undefined method `oid' for nil:NilClass`

**Comment by gettalong** [comment]
@pashamur Could you provide this PDF so that I can investigate the problem?

**Comment by matpowel** [comment]
@gettalong is this at all related to another error we get regularly like this? It seems to also be a corrupt PDF.

```Validation error for (33,0): Required field Parent is not set```

Unfortunately, I've only seen it happen on sensitive financial information so I can't send you the PDF :( 

For now, for both of these errors we have to catch the exceptions and send the document off for re-conversion back to PDF, but it would awesome if HexaPDF could handle such things if at all possible.

**Comment by pashamur** [comment]
@gettalong Unfortunately I'm not allowed to share the PDFs I was testing on :( 
(I ended up using PDFParser/poppler utils for my use cases)

**Comment by gettalong** [comment]
@matpowel No, I don't think so This issue is about parse errors due to corrupted documents. The PDF you have seems to be parsable because you can actually do something with it. However, there seems to be a structural problem with it, the `Parent` key is used e.g. by the PDF internal page tree structure. There are already mechanisms in place to repair invalid page trees during validation, so this may be something that is currently not caught. Can you run `hexapdf inspect YOUR.PDF p` on that file without error? Can you open and view that file in a PDF reader application?

@pashamur Thanks nonetheless!

**Comment by matpowel** [comment]
@gettalong  the error comes when trying a simple open and write without any modification. If we open the PDF in Mac Preview and re-export to PDF then it works fine. Previously we added code to try opening the doc and rescue the Malformed error when first importing the doc, now we added code to try writing to a tempfile and rescuing the Error. In both cases we send it off to ConvertAPI.com for reprocessing (we also use that to convert from Word etc) and then the PDF processes just fine.

**Comment by gettalong** [comment]
@matpowel If opening in Mac Preview works, I should be able to find the problem and make it work in HexaPDF, given you can provide the PDF file.

**Comment by gettalong** [comment]
@matpowel Could you try the following example scripts with your PDF file and the latest commit from the devel branch:

Prints the number of revisions and does a simple open/write:
~~~ ruby
require 'hexapdf'
HexaPDF::Document.open(ARGV[0]) do |doc|
  puts doc.revisions.size
  doc.write('test1.pdf', validate: true)
end
~~~

Simple open/write but without validations:
~~~ ruby
require 'hexapdf'
HexaPDF::Document.open(ARGV[0]) do |doc|
  doc.write('test2.pdf', validate: false)
end
~~~

Validating the document but only the current revision:
~~~ ruby
require 'hexapdf'
HexaPDF::Document.open(ARGV[0]) do |doc|
  each(only_current: true, only_loaded: false) do |obj|
    obj.validate(auto_correct: true) do |obj, msg, correctable|
      next if correctable
      raise HexaPDF::Error, "Validation error for (#{obj.oid},#{obj.gen}): #{msg}"
    end
  end
end
~~~

The problem might be that the PDF file contains multiple revisions and a prior revision is either invalid or the validation routine doesn't handle PDF with multiple revisions correctly.

**Comment by gettalong** [comment]
@matpowel I will close this issue since the original problem is solved with the next release. If you can try out the scripts in https://github.com/gettalong/hexapdf/issues/51#issuecomment-724119150 with your PDF file and one of them fails, please open a new issue - thank you!

**Comment by efatsi** [comment]
👏 👏 👏 👏

**Comment by matpowel** [comment]
Thanks @gettalong, we'll test over the coming weeks

**Comment by matpowel** [comment]
@gettalong we finally got to testing, this is now confirmed fixed for our test suite of PDFs in version 0.13.0 so we upgraded production now. Thank you!

**Comment by gettalong** [comment]
@matpowel Great to hear and thank you for your feedback!
